File: apps/dashboard/.claude.backup.2026-05-13T05-08-54/scripts/spec-link.js
#!/usr/bin/env node
'use strict';
/**
 * SPEC-LINK: Links a child spec to a parent spec (Wave 7 — parent/child hierarchy).
 *
 * CLI:
 *   node spec-link.js --parent <epic> --child <sub1> --reason "<why split>"
 *
 * Exported:
 *   linkSpec({ parent, child, reason, cwd? }) → boolean
 *
 * Effects:
 *   1. Emits event { event: "spec.link", payload: { parent, child, reason } } to harness log.
 *   2. Updates .pipeline-states/{parent}.json: adds child to children_specs (idempotent).
 *   3. Updates .pipeline-states/{child}.json: sets parent_spec (creates placeholder if missing).
 *
 * Guards:
 *   - Fail-open: exits 0 on any error (warns to stderr).
 *   - Node built-ins only. No npm deps.
 *   - Idempotent: re-linking same parent/child is a no-op.
 *
 * @version 1.0.0
 */

const fs = require('fs');
const path = require('path');

let harnessEvent = null;
try {
  harnessEvent = require('./../hooks/_lib/harness-event.js');
} catch (_) {}

/**
 * Read a pipeline-state file, returning parsed object or null.
 */
function readState(filePath) {
  try {
    if (!fs.existsSync(filePath)) return null;
    return JSON.parse(fs.readFileSync(filePath, 'utf8'));
  } catch (_) {
    return null;
  }
}

/**
 * Write a pipeline-state file safely. Returns true on success.
 */
function writeState(filePath, obj) {
  try {
    const dir = path.dirname(filePath);
    if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
    fs.writeFileSync(filePath, JSON.stringify(obj, null, 2) + '\n', 'utf8');
    return true;
  } catch (err) {
    process.stderr.write(`[spec-link] warn: could not write ${filePath}: ${err.message}\n`);
    return false;
  }
}

/**
 * Core logic. Returns true if all effects succeeded (partial success also returns true
 * as long as the event was emitted — fail-open design).
 *
 * @param {object} opts
 *   - parent  {string}  Epic spec name
 *   - child   {string}  Child spec name
 *   - reason  {string}  Why split (recorded in event)
 *   - cwd     {string}  Project root (default: process.cwd())
 * @returns {boolean}
 */
function linkSpec(opts) {
  const options = opts || {};
  const parent = typeof options.parent === 'string' ? options.parent.trim() : '';
  const child = typeof options.child === 'string' ? options.child.trim() : '';
  const reason = typeof options.reason === 'string' ? options.reason.trim() : '';
  const cwd = typeof options.cwd === 'string' ? options.cwd : process.cwd();

  if (!parent || !child) {
    process.stderr.write('[spec-link] warn: --parent and --child are required\n');
    return false;
  }

  // ── 1. Emit spec.link event ──────────────────────────────────────────────────
  let emitted = false;
  try {
    if (harnessEvent && typeof harnessEvent.emit === 'function') {
      emitted = harnessEvent.emit(
        'spec.link',
        { parent, child, reason },
        { cwd, actor: { kind: 'script', id: 'spec-link' } }
      );
    }
  } catch (_) {}

  // ── 2. Update parent pipeline-state ─────────────────────────────────────────
  const statesDir = path.join(cwd, '.claude', '.pipeline-states');
  const parentFile = path.join(statesDir, parent + '.json');
  let parentState = readState(parentFile);

  if (parentState === null) {
    // Parent state doesn't exist — create minimal placeholder
    parentState = {
      spec: parent,
      parent_spec: null,
      children_specs: [],
    };
  }

  // Ensure Wave 7 fields exist with safe defaults
  if (!parentState.parent_spec && parentState.parent_spec !== null) parentState.parent_spec = null;
  if (!Array.isArray(parentState.children_specs)) parentState.children_specs = [];

  // Idempotent: only add if not already present
  if (!parentState.children_specs.includes(child)) {
    parentState.children_specs.push(child);
  }

  writeState(parentFile, parentState);

  // ── 3. Update child pipeline-state ──────────────────────────────────────────
  const childFile = path.join(statesDir, child + '.json');
  let childState = readState(childFile);

  if (childState === null) {
    // Create minimal placeholder for child
    childState = {
      spec: child,
      parent_spec: parent,
      children_specs: [],
    };
  } else {
    if (!Array.isArray(childState.children_specs)) childState.children_specs = [];
    childState.parent_spec = parent;
  }

  writeState(childFile, childState);

  return true;
}

module.exports = { linkSpec };

// ── CLI ───────────────────────────────────────────────────────────────────────
if (require.main === module) {
  (function () {
    try {
      const args = process.argv.slice(2);

      function getArg(name) {
        const idx = args.indexOf('--' + name);
        return idx >= 0 ? args[idx + 1] : null;
      }

      const parent = getArg('parent');
      const child = getArg('child');
      const reason = getArg('reason') || '';
      const cwd = getArg('cwd') || process.cwd();

      if (!parent || !child) {
        process.stderr.write('Usage: node spec-link.js --parent <epic> --child <sub> --reason "<text>" [--cwd <path>]\n');
        process.exit(0);
      }

      const ok = linkSpec({ parent, child, reason, cwd });
      if (ok) {
        process.stdout.write(JSON.stringify({ ok: true, parent, child, reason }) + '\n');
      } else {
        process.stdout.write(JSON.stringify({ ok: false, parent, child, reason }) + '\n');
      }
    } catch (err) {
      process.stderr.write(`[spec-link] error: ${err.message}\n`);
    }
    process.exit(0);
  })();
}

File: apps/rt/src/run/spec_link.rs
//! `mustard-rt run spec-link` — a port of `scripts/spec-link.js`.
//!
//! Links a child spec to a parent spec (parent/child epic hierarchy):
//!
//! 1. Emits a `spec.link` harness event.
//! 2. Updates `.pipeline-states/{parent}.json`: adds the child to
//!    `children_specs` (idempotent).
//! 3. Updates `.pipeline-states/{child}.json`: sets `parent_spec` (creating a
//!    placeholder when absent).
//!
//! Port note: the JS version shelled to `_lib/harness-event.js` to emit the
//! event. This port appends the event directly through `mustard_core`.

use crate::run::env::session_id;
use crate::util::now_iso8601;
use mustard_core::fs;
use mustard_core::store::event_store::EventSink;
use mustard_core::store::sqlite_store::SqliteEventStore;
use mustard_core::model::event::{Actor, ActorKind, HarnessEvent, SCHEMA_VERSION};
use serde_json::{json, Value};
use std::path::Path;

/// Read a pipeline-state file, returning `None` on any error.
fn read_state(path: &Path) -> Option<Value> {
    let text = fs::read_to_string(path).ok()?;
    serde_json::from_str(&text).ok()
}

/// Write a pipeline-state file (pretty JSON + trailing newline). Fail-soft.
fn write_state(path: &Path, value: &Value) -> bool {
    match serde_json::to_string_pretty(value) {
        Ok(text) => fs::write_atomic(path, format!("{text}\n").as_bytes()).is_ok(),
        Err(_) => false,
    }
}

/// Emit a `spec.link` harness event. Best-effort.
///
/// **Spec attribution:** the event is attributed to the *child* spec — that is
/// the spec receiving the link (a follow-up, sub-feature, or addendum). The
/// parent shows up in the payload (`parent`, `reason`) so projections that
/// walk the lineage still have both names. Pre-2026-05-20 this event left
/// `spec = NULL`, which made `spec.link` rows invisible to projections that
/// filter by spec slug.
fn emit_link_event(cwd: &Path, parent: &str, child: &str, reason: &str) {
    let ev = HarnessEvent {
        v: SCHEMA_VERSION,
        ts: now_iso8601(),
        session_id: session_id(),
        wave: 0,
        actor: Actor {
            kind: ActorKind::Cli,
            id: Some("spec-link".to_string()),
            actor_type: None,
        },
        event: "spec.link".to_string(),
        payload: json!({ "parent": parent, "child": child, "reason": reason }),
        spec: Some(child.to_string()),
    };
    let _ = SqliteEventStore::for_project(cwd).and_then(|store| store.append(&ev));
}

/// Core link logic. Returns `true` when the link was applied (fail-open).
fn link_spec(cwd: &Path, parent: &str, child: &str, reason: &str) -> bool {
    let parent = parent.trim();
    let child = child.trim();
    if parent.is_empty() || child.is_empty() {
        eprintln!("[spec-link] warn: --parent and --child are required");
        return false;
    }

    emit_link_event(cwd, parent, child, reason);

    let states_dir = cwd.join(".claude").join(".pipeline-states");

    // Parent state — append the child to `children_specs` idempotently.
    let parent_file = states_dir.join(format!("{parent}.json"));
    let mut parent_state = read_state(&parent_file).unwrap_or_else(|| {
        json!({ "spec": parent, "parent_spec": Value::Null, "children_specs": [] })
    });
    if let Some(obj) = parent_state.as_object_mut() {
        if !obj.contains_key("parent_spec") {
            obj.insert("parent_spec".to_string(), Value::Null);
        }
        let children = obj
            .entry("children_specs")
            .or_insert_with(|| json!([]));
        if !children.is_array() {
            *children = json!([]);
        }
        if let Some(arr) = children.as_array_mut() {
            let present = arr.iter().any(|v| v.as_str() == Some(child));
            if !present {
                arr.push(json!(child));
            }
        }
    }
    write_state(&parent_file, &parent_state);

    // Child state — set `parent_spec`.
    let child_file = states_dir.join(format!("{child}.json"));
    let mut child_state = read_state(&child_file).unwrap_or_else(|| {
        json!({ "spec": child, "parent_spec": parent, "children_specs": [] })
    });
    if let Some(obj) = child_state.as_object_mut() {
        if !obj.get("children_specs").is_some_and(Value::is_array) {
            obj.insert("children_specs".to_string(), json!([]));
        }
        obj.insert("parent_spec".to_string(), json!(parent));
    }
    write_state(&child_file, &child_state);

    true
}

/// Dispatch `mustard-rt run spec-link`.
pub fn run(parent: Option<&str>, child: Option<&str>, reason: Option<&str>) {
    let cwd = std::env::current_dir().unwrap_or_else(|_| Path::new(".").to_path_buf());
    let (Some(parent), Some(child)) = (parent, child) else {
        eprintln!("Usage: spec-link --parent <epic> --child <sub> --reason \"<text>\"");
        return;
    };
    let reason = reason.unwrap_or("");
    let ok = link_spec(&cwd, parent, child, reason);
    println!(
        "{}",
        json!({ "ok": ok, "parent": parent, "child": child, "reason": reason })
    );
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::tempdir;

    #[test]
    fn link_creates_and_updates_states() {
        let dir = tempdir().unwrap();
        assert!(link_spec(dir.path(), "epic", "child-1", "split"));
        let states = dir.path().join(".claude").join(".pipeline-states");
        let parent = read_state(&states.join("epic.json")).unwrap();
        assert_eq!(
            parent["children_specs"],
            json!(["child-1"])
        );
        let child = read_state(&states.join("child-1.json")).unwrap();
        assert_eq!(child["parent_spec"], json!("epic"));
    }

    #[test]
    fn link_is_idempotent() {
        let dir = tempdir().unwrap();
        link_spec(dir.path(), "epic", "child-1", "split");
        link_spec(dir.path(), "epic", "child-1", "split");
        let states = dir.path().join(".claude").join(".pipeline-states");
        let parent = read_state(&states.join("epic.json")).unwrap();
        assert_eq!(parent["children_specs"].as_array().unwrap().len(), 1);
    }

    #[test]
    fn link_rejects_empty_args() {
        let dir = tempdir().unwrap();
        assert!(!link_spec(dir.path(), "", "child", "r"));
    }
}
